lib/TraackrApi/TraackrApiObject.php

```diff
@@ -2,256 +2,299 @@
+
+    // Prepare parameters before any GET or POST call.
+    // For now any pass-thru parameter passed as a true or false boolean
+    // is converted to a string since that's what the API expects
+    private function prepareParameters($params)
+    {
+        foreach ($params as $key => $value) {
+            if ($params[$key] === true) {
+                $params[$key] = 'true';
+            }
 
-   } // End constructor
+            if ($params[$key] === false) {
+                $params[$key] = 'false';
+            }
+        }
 
-   protected function checkRequiredParams($params, $fields) {
+        return $params;
+    }
 
-      foreach ($fields as $f) {
+    private function call($decode)
+    {
+        // Prep headers
+        curl_setopt(
+            $this->curl,
+            CURLOPT_HTTPHEADER,
+            array_merge($this->curl_headers, TraackrApi::getExtraHeaders())
+        );
 
-         // empty(false) returns true so need extra test for that
-         if ( empty($params[$f]) && !(isset($params[$f]) && is_bool($params[$f])) ) {
-            throw new MissingParameterException('Missing parameter: '.$f);
-         }
+        // Make the call!
+        $curl_exec = curl_exec($this->curl);
 
-      } // End fields loop
+        $logger = TraackrAPI::getLogger();
 
-   } // End function checkRequiredParams()
+        if ($curl_exec === false) {
+            $info = curl_getinfo($this->curl);
+            $message = 'API call failed (' . $info['url'] . '): ' . curl_error($this->curl);
 
-   protected function addCustomerKey(&$params) {
+            $logger->error($message);
 
-      $key = TraackrApi::getCustomerKey();
-      if ( !empty($key) && empty($params[PARAM_CUSTOMER_KEY]) ) {
-         $params[PARAM_CUSTOMER_KEY] = $key;
-      }
-      return $params;
+            throw new TraackrApiException($message);
+        }
 
-   } // End function addCustomerKey()
+        if (null === $curl_exec) {
+            $message = 'API call failed. Response was null.';
 
-   /*
-    * Make best attempt at converting booleans.
-    * Boolean type should be passed to the API but this function will also
-    * handle their string representation ('true' and 'false')
-    */
-   protected function convertBool($params, $key){
+            $logger->error($message);
 
-      // Does key even exists?
-      if ( !isset($params[$key]) ) return 'false';
+            throw new TraackrApiException($message);
+        }
 
-      $bool = $params[$key];
+        $httpcode = curl_getinfo($this->curl, CURLINFO_HTTP_CODE);
 
-      if ( is_bool($bool) ) return $bool ? 'true' : 'false';
+        if ($httpcode != '200') {
+            $info = curl_getinfo($this->curl);
 
-      if ( strtolower($bool) === 'true' ) return 'true';
+            if ($httpcode == '400') {
+                // Let's try to see if it's a bad customer key
+                if ($curl_exec === 'Customer key not found') {
+                    $message = 'Invalid Customer Key (HTTP 400)';
```

Comment:
Not sure why the diff is such a hot mess, but this is the pattern of the logging used. Capture the overall message and use that message in the logger call and exception thrown.

Reply:
So functionally nothing changes, but we now have error messages that are consistent between the exception thrown and what the logger records.... am I understanding this correctly?

Reply:
You are understanding this correctly.